src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -342,15 +357,22 @@ async def build_pex(
         # TODO(#9560): consider validating that these platforms are valid with the interpreter
         #  constraints.
         argv.extend(request.platforms.generate_pex_arg_list())
-    else:
-        argv.extend(request.interpreter_constraints.generate_pex_arg_list())
+    elif request.python:
+        python = request.python
+    elif request.internal_only:
         # NB: If it's an internal_only PEX, we do our own lookup of the interpreter based on the
         # interpreter constraints, and then will run the PEX with that specific interpreter. We
         # will have already validated that there were no platforms.
-        if request.internal_only:
-            python = await Get(
-                PythonExecutable, InterpreterConstraints, request.interpreter_constraints
-            )
+        python = await Get(
+            PythonExecutable, InterpreterConstraints, request.interpreter_constraints
+        )
+    else:
+        # `--interpreter-constraint` options are mutually exclusive with the `--python` option,
+        # so we only specify them if we have not already located a concrete Python.
+        argv.extend(request.interpreter_constraints.generate_pex_arg_list())
```

Comment:
NB: This is a significant change. The `local_only` option was _not_ having the effect of overriding the interpreter constraints: instead, it was using the selected `PythonExecutable` to run PEX, but PEX would still use the interpreter constraints that it was given to find a match from the `--python-path`.

By passing `--python` (which is mutually exclusive with `--interpreter-constraint`), we now actually force a particular python for `local_only=True`... which was also necessary to fix this bug.

Reply:
I suspect that will result in a performance boost

Reply:
By omitting the ICs for internal-only PEXes, they are created with the desired interpreter but otherwise unconstrained at runtime. I.E: There is no trace left in the PEX file output (PEX-INFO in particular) of the ICs. As such an internal_only PEX file built with IC "==3.7.** will produce the same PEX file as one built with IC "==3.8.*" if the PEX requirements happen to be agnostic to that difference.

Concretely, if we build a setuptools PEX with IC "==3.7.*" 1st, then we build one with IC "==3.8.*" we get the same output PEX which leads downstream rules, notably VenvPex reliant ones, missing an encoding of the IC. For a normal internal_only PEX file, we have the same problem, but we run them via `python pex_file ...` where the `python` is the one picked from the otherwise erased ICs. In the VenvPex case, we can do no such thing since a venv always (must) uses its interpreter.

I'm not sure how best to deal with this yet, but planting a flag.

Reply:
#12635